drivers: mcux: Flexcomm uart async api

decsny opened the pull request:
This PR implements the UART Async API for the mcux flexcomm uart driver.

depends on #57719

Commit:
drivers: uart_mcux_flexcomm: Simplify init macros

The flexcomm uart mcux driver init macros have a lot
of unnecessary and confusing redundancy, so let's
simplify them.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: uart_mcux_flexcomm: Clean up

Clean up slashes in init macro to be less messy

Update copyright to reflect recent code changes by NXP

Update outdated driver description comment

Change irq driven api variables naming to be more specific

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: uart_mcux_flexcomm: Add ASYNC API

Support ASYNC API on Flexcomm UART

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
boards: mimxrtXXX: Add UART DMAs

Add DMA phandles to UART nodes in RT 3 digit EVK
dts files now that the UART ASYNC API is supported
in the mcux flexcomm driver.

Also add the DMA phandles to other platforms with
flexcomm uart and lpc dma enabled by default for
compatibility

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

dleach02 commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -346,6 +378,577 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	unsigned int key = irq_lock();
+
+	/* Getting DMA status to tell if channel is busy or not set up */
+	struct dma_status dma_status;
```
I guess the Zephyr coding guidelines require the name of the variable to be different from the structure name.

decsny replied:
it seems to be okay with just `status`

teburd reviewed (dismissed):
dma related changes look good, I can't speak to the uart changes

decsny commented:
need to remove those dma changes actually, they are old commits, the updated changes were merged in #57719

Commit:
tests: uart_async_api: Support RTXXX

Support mimxrt595_evk_cm33 and mimxrt685_evk_cm33
on the uart_async_api test with a board overlay

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	ret = dma_get_status(config->tx_dma.dev, config->tx_dma.channel, &status);
+
+	if (ret < 0) {
+#ifdef CONFIG_DMA_MCUX_LPC
```
The custom behavior here based on which DMA driver is being used concerns me. I assume that the first time we attempt to transmit using this channel it would be unconfigured, and  we would get the `-EACCES` return code.


Rather than using this driver specific behavior, could you instead track if the TX is busy using something like the state of `data->tx_data.xfer_len`, or another variable? It seems based on the implementation here that `data->tx_data.xfer_len` would only be 0 when a transfer is not active (you of course would need to initialized `data->tx_data.xfer_len = 0` in the driver init for this to work)

decsny replied:
Well I think maybe not because this channel can also be used for DMIC requests and memory to memory, and I don't want to make any assumptions about how a user is or isn't writing their application regarding reconfiguring and re-using DMA channels at runtime. And you can't just check if tx_data.xfer_len is 0 && ret = 0, because we need to differentiate the -eacces code from the -ebusy

Also, this is more a problem with that DMA driver, maybe it shouldn't be returning -EACCES, maybe it should just say "not busy" if a channel is not set up to do anything. The EACCES code is not part of the dma api, its a custom error code, it was added because of something like the dma driver interpreting the channel argument as a "virtual channel" in the dma driver and lots of k_mallocs and  garbage values, I don't remember those details too well now, it's a bigger problem with that DMA driver imo that it shouldnt be using that virtual channel scheme.

decsny replied:
I think it would be better to fix the bug in the DMA driver

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	}
+
+	/* Enable interrupt for when TX fifo is empty (all data transmitted) */
+	config->base->FIFOINTENSET |= USART_FIFOINTENSET_TXLVL_MASK;
```
Two comments here:
1. Could we use the TX IDLE interrupt to signal all data is transmitted? It is not clear to me from the hardware RM if the TXLVL interrupt would fire before the transfer is complete when more than 16 bytes are sent (as in this case the TX FIFO would fill, drain, and then the DMA would fill it again). As a side benefit, this way we can configure the TXLVL watermark higher than 0, if we want to set up the USART this way in the future.
2. Why not use the HAL API here?

decsny replied:
how would TX idle be different from  TXLVL, does the transmitter not idle when the fifo is empty and the uart has no data to send?

danieldegrasse replied:
The reference manual for the USART made it unclear to me how TXLVL would function when using DMA. In the case of DMA, it seems that when the TX FIFO is drained the TXLVL interrupt will be asserted (or the DMA request, if the interrupt signal is disabled. My concern was that the TX FIFO might become drained in some cases prior to the DMA writing new data, resulting in multiple TX done events being sent (one per interrupt) however, I have tested for this issue using PR locally, and that does not seem to occur. So I'm going to go ahead and resolve this comment.

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	ret = dma_get_status(config->rx_dma.dev, config->rx_dma.channel, &status);
+
+	if (ret < 0) {
+#ifdef CONFIG_DMA_MCUX_LPC
```
Same comment as above about DMA driver specific behavior, but in this case using `data->rx_data.xfer_buf`

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	 * This is only needed if using a timeout.
+	 */
+	if (timeout != SYS_FOREVER_US) {
+		config->base->INTENSET |= USART_INTENSET_STARTEN_MASK;
```
Can we use the HAL API here?

decsny replied:
In general I am trying to avoid using the HAL API, we can talk about this offline because I understand at least 3 points for using it and 3 points against using it but I think it's not necessary to have this discussion here

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+}
+
+#if defined(CONFIG_SOC_SERIES_IMX_RT5XX) || defined(CONFIG_SOC_SERIES_IMX_RT6XX)
+static uint32_t fc_uart_calc_inmux_connection(uint8_t channel, DMA_Type *base)
```
A comment documenting that this function calculates the required inputmux connection for a given channel would probably be useful for future developers

decsny replied:
that's the name of the function, I can un-abbreviate

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+			config->rx_timeout_func);
+
+	/* TX FIFO interrupt; configure to interrupt when empty */
+	config->base->FIFOTRIG &= ~USART_FIFOSTAT_TXLVL_MASK;
```
This is already set in the USART_Init call, by the `txWatermark` field in the configuration structure (which defaults to 0)

decsny replied:
I know, I specifically only included this line for clarity. Do you think it should be removed

danieldegrasse replied:
Ideally yes, remove it and add a comment that clearing the FIFOTRIG register is not required.

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+						K_USEC(data->rx_data.timeout));
+			}
+
+			/* W1C start bit status bit */
```
"Write one to clear" (please be more verbose here). Also, isn't `USART_STAT_START_MASK` defined as `0x1000`? So won't the bitmask here make this effectively a NOP, or am I missing something?

decsny replied:
yes it is completely redundant, I will change it

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+
+		/* Handle TX interrupt (TXLVL = 0)
+		 * TX interrupt should happen when transfer is complete
+		 * because DMA filling TX fifo is faster than transmitter rate
```
Just to understand this comment a bit better here- is the reason that using TXLVL=0 as the interrupt condition (rather the the TXIDLE bit) safe because the DMA engine will refill the FIFO before the USART can write enough data to empty it? I want to be sure that using TXLVL=0 as the condition does not result in the driver indicating to the user that the transfer is complete before all data was sent (especially in cases where we transmit more than 16 bytes, and the FIFO would be full and thus stall the DMA)

decsny replied:
yes

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+			&& data->async_callback) {
+
+			/* Disable interrupt */
+			config->base->FIFOINTENCLR |= USART_FIFOINTENCLR_TXLVL_MASK;
```
Do we need `|=` here? This looks like a register where we can just write directly, as it seems all the bits in are read as zero

decsny replied:
I think it would look strange/confusing in the code to have just `=` when writing to a register to clear a specific bit, this is the common practice

danieldegrasse replied:
Writing directly to a `SET` register is common practice- it avoids an unnecessary register read, and is one of the reasons that a `SET` register exists at all.
```
grep -r "_SET = ." --include="*.c"
drivers/timer/mchp_xec_rtos_timer.c:    ECIA_XEC_REGS->GIRQ[girq - 8].EN_SET = BIT(bitpos);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:              regs->BLK_EN_SET = BIT(girq_num);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      regs->GIRQ[girq_num - MCHP_FIRST_GIRQ].EN_SET = BIT(src_bit_pos);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      regs->GIRQ[girq_num - MCHP_FIRST_GIRQ].EN_SET = bitmap;
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      regs->GIRQ[girq - MCHP_FIRST_GIRQ].EN_SET = BIT(src);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      ecia->BLK_EN_SET = MCHP_ECIA_AGGR_BITMAP;
soc/arm/microchip_mec/mec1501/device_power.c:   GIRQ22_REGS->EN_SET = (1ul << 9);
soc/arm/microchip_mec/mec1501/soc.c:    ECIA_REGS->BLK_EN_SET = MCHP_ECIA_AGGR_BITMAP;
soc/arm/microchip_mec/mec172x/device_power.c:   regs->GIRQ22.EN_SET = MCHP_ESPI_WK_CLK_GIRQ_BIT;
soc/arm/microchip_mec/mec172x/device_power.c:   regs->GIRQ21.EN_SET = MCHP_KEYSCAN_GIRQ_BIT;
soc/arm/microchip_mec/mec172x/device_power.c:   regs->GIRQ21.EN_SET = MCHP_PS2_0_PORT0B_WK_GIRQ_BIT;
soc/arm/nxp_imx/rt/power_rt10xx.c:              XTALOSC24M->LOWPWR_CTRL_SET = XTALOSC24M_LOWPWR_CTRL_LPBG_SEL_MASK;
soc/arm/nxp_imx/rt/power_rt10xx.c:              PMU->MISC0_SET = PMU_MISC0_REFTOP_PWD_MASK;
soc/arm/nxp_imx/rt/lpm_rt1064.c:        CCM_ANALOG->PLL_USB1_SET = CCM_ANALOG_PLL_USB1_BYPASS_MASK;
soc/arm/nxp_imx/rt/lpm_rt1064.c:        CCM_ANALOG->PLL_USB1_SET = CCM_ANALOG_PLL_USB1_ENABLE_MASK;
soc/arm/nxp_imx/rt/soc_rt11xx.c:        CCM->GPR_PRIVATE1_SET = 0x1;
soc/arm/nxp_imx/rt5xx/flash_clock_setup.c:      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI0_OTFAD_CLK_MASK;
soc/arm/nxp_imx/rt5xx/flash_clock_setup.c:                      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI0_OTFAD_CLK_MASK;
soc/arm/nxp_imx/rt5xx/flash_clock_setup.c:                      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI1_CLK_MASK;
soc/arm/nxp_imx/rt6xx/flash_clock_setup.c:                      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI_OTFAD_CLK_MASK;
```

decsny replied:
ok I guess the same is true for the set register then too?

decsny commented:
@danieldegrasse can you check the changes, I didn't rebase so that they are easy to see from the compare button

mmahadevan108 commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,640 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+		 */
+		if (config->base->INTSTAT & USART_INTSTAT_START_MASK
+			&& data->async_callback) {
+
```
Don't need `&& data->async_callback` as the check is already made earlier.

decsny replied:
good catch I left those by accident

Commit:
drivers: dma_mcux_lpc: Fix busy status

Driver should just say the channel is not busy if
it is not setup rather than returning an error.

Also, change the channel index to int8_t rather
than uint32_t since it is being assigned negative
values and that width is more appropriate.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -474,7 +474,7 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
 {
 	struct dma_mcux_lpc_dma_data *dev_data = dev->data;
-	uint32_t virtual_channel = dev_data->channel_index[channel];
+	int8_t virtual_channel = dev_data->channel_index[channel];
```
Please also change `channel_index` in `dma_mcux_lpc_dma_data` structure.

mmahadevan108 commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,612 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	return ret;
+}
+
+static void flexcomm_uart_rx_update(const struct device *dev)
```
Should we call this `mcux_flexcomm_uart_rx_update` to be consistent with the naming of functions in this file.

mmahadevan108 commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -342,6 +379,612 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+#endif /* RT 3-digit */
+
+
+static int flexcomm_uart_async_init(const struct device *dev)
```
Call this `mcux_flexcomm_uart_async_init`?
